planetary-variables/soil-water-content/soil-water-content-flash-drought/raw.js

```diff
@@ -0,0 +1,82 @@
+//VERSION=3
+
+const nDaysBackwardAverage = 14; // The number of days that is used to calculate the swc backward average
+const nDaysPrevious = 28; // The number of days to look back to compare the current value with the previous value
+const scaleFactor = 1000; // The scale factor for the SWC values
+const droughtThreshold = 0.15; // The SWC value under which conditions are considered drought-like
+const differenceThreshold = -0.12; // Threshold signifying a rapid drop in SWC from the previous average value
+
+function setup() {
+  return {
+    input: ["SWC", "dataMask"],
+    output: { bands: 1 },
+    mosaicking: "ORBIT",
+  };
+}
+
+const daysToLoad = nDaysBackwardAverage + nDaysPrevious; // The number of days looking back to load data for
+
+function preProcessScenes(collections) {
+  const millisecondsBack = daysToLoad * 24 * 3600 * 1000;
+
+  collections.scenes.orbits = collections.scenes.orbits.filter(function (
+    orbit
+  ) {
+    const orbitDateFrom = new Date(orbit.dateFrom);
+    return (
+      orbitDateFrom.getTime() >= collections.to.getTime() - millisecondsBack
+    );
+  });
+  return collections;
+}
+
+function getMeanSWCValue(swc, dataMask) {
+  // Get the sum of all SWC values
+  let validDateCount = 0;
+  let sum = 0;
+  for (let i = 0; i < swc.length; i++) {
+    if (dataMask[i]) {
+      sum += swc[i];
+      validDateCount += 1;
+    }
+  }
+
+  // Calculate the mean SWC value
+  let meanSWCValue = NaN;
+  if (validDateCount > 0) {
+    meanSWCValue = sum / validDateCount;
+  }
+
+  return meanSWCValue;
+}
+
+function evaluatePixel(samples) {
+  // When there are no dates, return no data
+  if (samples.length == 0) return [NaN];
+  // When the search interval of the request body is too short to perform the calculation, return no data
+  if (samples.length < daysToLoad) return [NaN];
+
+  // Extract SWC values and dataMask
+  const swc = samples
+    .slice(0, nDaysBackwardAverage)
+    .map((sample) => sample.SWC / scaleFactor);
+  const dataMask = samples
+    .slice(0, nDaysBackwardAverage)
+    .map((sample) => sample.dataMask);
+  const swcPrevious = samples
+    .slice(nDaysPrevious, nDaysBackwardAverage + nDaysPrevious)
+    .map((sample) => sample.SWC / scaleFactor);
+  const dataMaskPrevious = samples
+    .slice(nDaysPrevious, nDaysBackwardAverage + nDaysPrevious)
+    .map((sample) => sample.dataMask);
+
+  // Calculate mean SWC value
+  const meanSWC = getMeanSWCValue(swc, dataMask);
+  const meanSWCPrevious = getMeanSWCValue(swcPrevious, dataMaskPrevious);
```

Comment:
In general it is better to try and avoid `.map()` calls since they can be slow. Instead we can loop through only once on the sample object directly
```suggestion
  const initialSamples = samples
    .slice(0, nDaysBackwardAverage)
  const previousSamples = samples
    .slice(nDaysPrevious, nDaysBackwardAverage + nDaysPrevious)

  // Calculate mean SWC value
  const meanSWC = getMeanSWCValue(initialSamples);
  const meanSWCPrevious = getMeanSWCValue(previousSamples);
```